update-testsuite.py

```diff
@@ -0,0 +1,203 @@
+    for path, _repo_path, dst in spec.list_tests():
+        shutil.copyfile(path, dst)
+
+    # Process all upstream repositories and proposals.
+    failed_merges = []
+    updated = []
+    for repo in repos:
+        # Repositories may not be mergable with the upstream spec repository in
+        # which case we skip them and print an informational message at the end.
+        repo.update()
+        if not repo.checkout_merge(spec):
+            failed_merges.append(repo)
+            continue
+
+        # Blow away this proposal's list of tests if it exists.
+        dstdir = f'proposals/{repo.dir}'
+        if os.path.isdir(dstdir):
+            shutil.rmtree(dstdir)
+
+        # For all tests in this proposal run a diff against the upstream
+        # revision. If the diff is non-empty then include this test by copying
+        # it to its destination.
+        for path, repo_path, dst in spec.list_tests():
+            upstream = repo.upstream or spec
+
+            diff = repo.git('diff', upstream.rev, repo.merged_rev, '--', repo_path, quiet=True)
+            if len(diff) == 0:
+                continue
+
+            dst = os.path.join(dstdir, dst)
+            os.makedirs(os.path.dirname(dst), exist_ok=True)
+            shutil.copyfile(path, dst)
+
+        # If anything changed, then this is an updated proposal, and take note
+        # of that for later.
+        git('add', dstdir)
+        status = git('status', '-s', dstdir)
+        if len(status) > 0:
+            updated.append(repo)
+
+    for repo in failed_merges:
+        print('!! failed to update:', repo.url())
+
+    # If anything was updated, make a commit message indicating as such.
+    if len(updated) == 0:
+        return
+    message = "Update repos:\n\n"
+    for repo in updated:
+        message += f"  {repo.dir}:\n"
+        message += f"    {repo.url()}/commit/{repo.rev}\n"
+    message += "\n"
+    message += "This change was automatically generated by `update-testsuite.py`"
+    git('commit', '-a', '-m', message)
+
+class Repo:
+    def __init__(self, repo, branch = 'main', upstream = None):
+        self.repo = repo
+        if branch == 'main':
+            self.dir = repo
+        else:
+            self.dir = branch
+        self.branch = branch
+        self.upstream = upstream
+
+    def __str__(self):
+        repo = f'WebAssembly/{self.repo}'
+        if self.branch != 'main':
+            repo += f'@{self.branch}'
+        return repo
+
+    # Fetch the latest revision from this repository and store the revision
+    # within `self` of what was found.
+    def update(self):
+        self.git('fetch', self.url(), self.branch)
+        self.rev = self.git('rev-parse', 'FETCH_HEAD', quiet=True)
+
+    def url(self):
+        return f'https://github.com/WebAssembly/{self.repo}'
+
+    # Helper to run a `git` command within the `repos/spec` repository.
+    def git(self, *args, **kwargs):
+        return git('-C', 'repos/spec', *args, **kwargs)
```

Comment:
This seems unrelated to any specific repo so maybe move it outside the class or at least mark it as `@staticmethod`